src/main/java/org/opensearch/security/action/apitokens/ApiTokenIndexHandler.java

```diff
@@ -0,0 +1,149 @@
+
+import com.google.common.collect.ImmutableMap;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import org.opensearch.action.admin.indices.create.CreateIndexRequest;
+import org.opensearch.action.index.IndexRequest;
+import org.opensearch.action.index.IndexResponse;
+import org.opensearch.action.search.SearchRequest;
+import org.opensearch.action.search.SearchResponse;
+import org.opensearch.client.Client;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.util.concurrent.ThreadContext;
+import org.opensearch.common.xcontent.XContentFactory;
+import org.opensearch.common.xcontent.XContentType;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.core.xcontent.DeprecationHandler;
+import org.opensearch.core.xcontent.NamedXContentRegistry;
+import org.opensearch.core.xcontent.ToXContent;
+import org.opensearch.core.xcontent.XContentBuilder;
+import org.opensearch.core.xcontent.XContentParser;
+import org.opensearch.index.query.QueryBuilders;
+import org.opensearch.index.reindex.BulkByScrollResponse;
+import org.opensearch.index.reindex.DeleteByQueryAction;
+import org.opensearch.index.reindex.DeleteByQueryRequest;
+import org.opensearch.search.SearchHit;
+import org.opensearch.search.builder.SearchSourceBuilder;
+import org.opensearch.security.support.ConfigConstants;
+
+import static org.opensearch.security.action.apitokens.ApiToken.NAME_FIELD;
+
+public class ApiTokenIndexHandler {
+
+    private final Client client;
+    private final ClusterService clusterService;
+    private static final Logger LOGGER = LogManager.getLogger(ApiTokenIndexHandler.class);
+
+    public ApiTokenIndexHandler(Client client, ClusterService clusterService) {
+        this.client = client;
+        this.clusterService = clusterService;
+    }
+
+    public String indexTokenPayload(ApiToken token) {
+        try (final ThreadContext.StoredContext ctx = client.threadPool().getThreadContext().stashContext()) {
+
+            XContentBuilder builder = XContentFactory.jsonBuilder();
+            String jsonString = token.toXContent(builder, ToXContent.EMPTY_PARAMS).toString();
+
+            IndexRequest request = new IndexRequest(ConfigConstants.OPENSEARCH_API_TOKENS_INDEX).source(jsonString, XContentType.JSON);
+
+            ActionListener<IndexResponse> irListener = ActionListener.wrap(idxResponse -> {
+                LOGGER.info("Created {} entry.", ConfigConstants.OPENSEARCH_API_TOKENS_INDEX);
+            }, (failResponse) -> {
+                LOGGER.error(failResponse.getMessage());
+                LOGGER.info("Failed to create {} entry.", ConfigConstants.OPENSEARCH_API_TOKENS_INDEX);
+            });
+
+            client.index(request, irListener);
+            return token.getName();
+
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+
+    }
+
+    public void deleteToken(String name) throws ApiTokenException {
```

Comment:
Any audit logging present on internal user endpoints should be applicable here as well. Same audit logging for create user or delete user should have an equivalent entry here.

Reply:
I added a hack to do so by preserving the user in the indexhandler. This seems to be taken care of by the abstractAPIaction, so I think we will need to consolidate/abstract this away to a common place. We can also handle the authorization of this API in that abstraction, since that is something else that needs to happen. For now confirmed that the hack is working and an audit log entry for api tokens actions is created.